### DISCUSSION TITLE
2024/01/08/live-reconfiguring-druby/

### DISCUSSION THREAD

**Discussion by giscus** [katafrakt/website_comments#3] [Announcements]
# 2024/01/08/live-reconfiguring-druby/

Using dRuby to achieve simple live reconfiguration ability, without the need for constant checking an external database or restart the application.

https://katafrakt.me/2024/01/08/live-reconfiguring-druby/

**Comment by vidarh** [comment]
Hi, it's awesome to see more people explore dRuby, as it's under-utilised, and the idea of e.g. exposing the ability to change settings at runtime like this or reading out data is great.

However, I just wanted to add a word of caution, and a few suggestions:

In your irb session, if you do "config.instance_eval { undef :instance_eval }" to remove the drb proxy object's instance_eval method, you can do "config.instance_eval('puts "this will run on the server'")" because drb will forward almost *anything* to the server and casually execute it. This **is** mentioned in the docs, but people often skim through docs.

Chapter 12 of the druby book goes into a bit of detail on the druby security model:
https://www.druby.org/sidruby/12-security-in-druby.html

A couple of concrete considerations for people:

 * Bind to UNIX sockets rather than TCP sockets unless you *need* to be able to connect from a remote machine.
 * Make sure the port you bind to isn't accessible from outside your network if you need it to be accessible over TCP
 * If you can, inherit your front object from BasicObject instead of Object to make it harder to "escape"
 * Don't think securing your front-object is enough - *any* object directly or indirectly reachable from your frontobject that isn't an "undumped" object (see the docs; "basic" objects like booleans, numbers, strings are generally ok; you can also make your classes undumped) results in a druby proxy object being returned, and becomes a potential way in.
 * Set $SAFE level as per the book above if you can.
 * Run your server in an unprivileged account or in a container/vm if you can, so compromising the process doesn't break the rest of your system.

This isn't meant to discourage you - I rely on dRuby every day (my text editor uses it to access its buffers) and love it - just wanted people to be aware that the power comes with big footguns.

**Reply by katafrakt** [reply]
Thank you for this very valuable comment! As you could probably take from my blog post, I'm mostly using dRuby as a toy, so great to hear from someone who's using it regularly. Much appreciated.

**Comment by MatheusRich** [comment]
Sounds like this could be used to flip Feature Flags 😅
